controllers/C_Document.class.php

```diff
@@ -795,6 +801,16 @@ public function retrieve_action(string $patient_id = null, $document_id, $as_fil
         if (file_exists($temp_url)) {
             $url = $temp_url;
         }
+        //fire a remote call to see if the file is stored somewhere else
+        $s3Key = explode("//", $temp_url); //split the url to get the s3 key
+        $retrieveOffsiteDocument = new PatientRetrieveOffsiteDocument("/" . $s3Key[1]);
+        $this->eventDispatcher->dispatch($retrieveOffsiteDocument, PatientRetrieveOffsiteDocument::REMOTE_DOCUMENT_LOCATION);
+        //this is for the s3 bucket module. If the file is not found locally, it will be found remotely
+        if ($retrieveOffsiteDocument->getPresignedUrl() != null) {
```

Comment:
@juggernautsei This is exactly what I wanted.  However, can we change the method name to be something like **getOffsiteUrl()** instead of _getPresignedUrl()_?  

As we'll have these events in core for a long time, someone may have an offsite url that doesn't use the nomenclature of 'presigning' like AWS does. Its a minor detail, but I think it will help with future maintainability of the event.  In your module the offsite url is presigned, but other modules may just be a straight up link on an external webserver.

At that point it looks good to me and I can merge it in.

Reply:
@adunsulag changes are deployed. Generic it is.